ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/recovery/AtomicActionRecoveryModule.java

```diff
@@ -214,16 +216,18 @@ private void processTransactionsStatus() {
                             _transactionType) != StateStatus.OS_UNKNOWN) {
                         RecoverAtomicAction rcvAtomicAction = doRecoverTransaction(currentUid);
 
-                        /*
-                         * hasFailedParticipants() relies on reportHeuristics being set to true
-                         * during replay_completion in RecoverAtomicAction
-                         */
-                        if (rcvAtomicAction.hasFailedParticipants() || rcvAtomicAction.hasPreparedParticipants()) {
-                            this.hasWorkLeftToDo = true;
-                        } else if (rcvAtomicAction.hasHeuristicParticipants()) {
-                            tsLogger.logger.tracef(
-                                    "AtomicActionRecoveryModule.processTransactionsStatus heuristic action {0} " +
-                                            "was ignored during the assessment of leftover work.", currentUid);
+                        if (isWaitForWorkLeftToDo && Objects.nonNull(rcvAtomicAction)) {
```

Comment:
Please can I ask if this additional check of `Objects.nonNull(rcvAtomicAction)` was needed before or was introduced through JBTM-3894 somehow?

I am going to approve the pull but if the need for it was not introduced by JBTM-3894 then I think we should raise a JBTM to investigate adding moving this check outside of the check of isWaitForWorkLeftToDo

Reply:
I think it should have been part of JBTM-3894.